### DISCUSSION TITLE
qt 使用软渲染方式时崩溃？ 发现是共享opengl 上下文为空这个是bug 吗？

### DISCUSSION THREAD

**Discussion by weijibin-tal** [Tencent/libpag#2643] [Q&A] [answered]
QApplication::setAttribute(Qt::AA_UseSoftwareOpenGL);

<img width="923" alt="Snipaste_2024-12-27_11-10-52" src="https://github.com/user-attachments/assets/86255b06-e826-470c-9e13-3699147b7af1" />

**Comment by weijibin-tal** [comment]
![Uploading Snipaste_2024-12-27_11-12-27.png…]()

**Reply by domchen** [reply, marked as answer]
TGFX 是一个纯 GPU 渲染的方案，你在 QT 平台原生使用 TGFX 就只能 GPU 渲染，不能软件渲染。现代的 UI 框架已经几乎不使用软件渲染了，性能会比较差，你的场景具体是什么？如果有不得不使用软件渲染的场景，可以使用SwiftShader这个库去用软件模拟 OpenGL 环境。参考 TGFX 在Linux 端的编译过程就行，但是渲染完的结果没法直接上屏，需要自己读取像素然后写入到QT的界面上。这个也是一个性能灾难。

**Reply by weijibin-tal** [reply]
我们是使用QWidget 上播放Pag 的动画，目前使用的方案是使用 pag::PAGSurface::MakeOffscreen  创建一个离屏渲染的 surface ，然后使用readPixels 方法  读取像素，然后paint 到 QWidget 上，目前遇到的问题：设置Qt::AA_UseSoftwareOpenGL  使用离屏渲染的方案 readPixels 读取的像素是错误的，然后尝试使用QQuickItem来生成surface，发现会崩溃。

**Reply by weijibin-tal** [reply]
使用的 pag 版本 4.4.1.0 是 自己编译的 支持qt 的pag库

**Reply by domchen** [reply]
PAG的问题请在libpag仓库提问。解决方式前面说过了，你想用软件渲染只能编译一个swiftshader的版本。
